### ISSUE TITLE
Cairo setSource() leaks memory

### ISSUE THREAD

**Issue by StefanSalewski** [StefanSalewski/gintro#168]
A better solution which works for drawingarea.nim example is

```nim
proc setSource*(cr: Context; source: Pattern) =
  let oldpattern = cairo_get_source(cr.impl)
  if oldpattern != nil:
    let o = cairo_pattern_get_user_data(oldpattern, NUDK)
    if o != nil:
      GC_unref(cast[Pattern](o))
  GC_ref(source)
  cairo_set_source(cr.impl, source.impl)
  discard cairo_pattern_set_user_data(source.impl, NUDK, cast[pointer](source), nil) # store the Nim proxy object
  discard cairo_set_user_data(cr.impl, NUDK, cast[pointer](source), gcuref) # when Context gets collected Pattern is unreffed
```

Unfortunately we have to do the oldpattern check also for the procs that do set RGBA patterns!

But at least it works, and maybe we can find even a better solution, maybe call the oldpattern code somehow by cairo lib when cairo pattern refcount drops to zero by use of cairo_pattern_set_user_data(). First tries failed for that.

**Comment by StefanSalewski** [comment]
The other idea was something like

```nim
proc newgcurefpattern(p: pointer) {.cdecl.} =
  let oldpattern = cast[ptr Pattern00](p)
  if oldpattern != nil:
    let o = cairo_pattern_get_user_data(oldpattern, NUDK)
    if o != nil:
      discard cairo_pattern_reference(oldpattern)
      GC_unref(cast[Pattern](o))
  
proc cairo_set_source*(cr: ptr Context00; source: ptr Pattern00) {.importc, libcairo.}
#
proc setSource*(cr: Context; source: Pattern) =
  GC_ref(source)
  cairo_set_source(cr.impl, source.impl) # cairo will unref the old pattern and may call newgcurefpattern
  discard cairo_pattern_set_user_data(source.impl, NUDK, cast[pointer](source.impl), newgcurefpattern) # store the Nim proxy object
  discard cairo_set_user_data(cr.impl, NUDK, cast[pointer](source), gcuref) # when Context gets collected Pattern is unreffed
  cairo_pattern_destroy(source.impl)
```

But the problem is, that newgcurefpattern() is called only if refcount of pattern reached zero, and then cairo_pattern_reference(oldpattern) can not be used again to keep pattern alive. So cairos set_usert_data() is less powerful as the gobject toggle references. So it seems that we have to use the first solution from above.

**Comment by StefanSalewski** [comment]
The funny fact is that the setting of plain colors as source does not change the actual pattern set before. In

```nim
proc setSource*(cr: Context; red, green, blue: float) =
  #tryReleasePat(cr.impl)
  echo cast[int](cairo_get_source(cr.impl))
  cairo_set_source_rgb(cr.impl, red.cdouble, green.cdouble, blue.cdouble)
  echo cast[int](cairo_get_source(cr.impl)), "++"
```

both echo statements prints the same value. So setting a plain color is basically a NOP for the pattern, and finally we do need only

```
proc tryReleasePat(crimpl: ptr Context00) =
  let oldpattern = cairo_get_source(crimpl)
  if oldpattern != nil:
    let o = cairo_pattern_get_user_data(oldpattern, NUDK)
    if o != nil:
      discard cairo_pattern_set_user_data(oldpattern, NUDK, nil, nil)
      GC_unref(cast[Pattern](o))

proc cairo_set_source*(cr: ptr Context00; source: ptr Pattern00) {.importc, libcairo.}
#
proc setSource*(cr: Context; source: Pattern) =
  GC_ref(source)
  tryReleasePat(cr.impl)
  #GC_ref(source)
  cairo_set_source(cr.impl, source.impl)
  discard cairo_pattern_set_user_data(source.impl, NUDK, cast[pointer](source), nil) # store reference to the Nim proxy object
  discard cairo_set_user_data(cr.impl, NUDK, cast[pointer](source), gcuref) # When Context gets collected Pattern is unreffed
```

If discard cairo_pattern_set_user_data(oldpattern, NUDK, nil, nil) is really necessary is not really clear yet. Maybe it is necessary to call GC_ref(source) before tryReleasePat(cr.impl) when we set the same pattern again?

**Comment by StefanSalewski** [comment]
Should be fixed in v0.9.4 now.
Will close this issue, please reopen if necessary.
